beacon-chain/rpc/subscriber/api/events/handler.go

```diff
@@ -0,0 +1,203 @@
+	es        *EventSystem
+	unsubOnce sync.Once
+}
+
+// Err returns a channel that is closed when unsubscribed.
+func (sub *Subscription) Err() <-chan error {
+	return sub.f.err
+}
+
+// Unsubscribe uninstalls the subscription from the event broadcast loop.
+func (sub *Subscription) Unsubscribe() {
+	sub.unsubOnce.Do(func() {
+	uninstallLoop:
+		for {
+			// write uninstall request and consume logs/hashes. This prevents
+			// the eventLoop broadcast method to deadlock when writing to the
+			// filter event channel while the subscription loop is waiting for
+			// this method to return (and thus not reading these events).
+			select {
+			case sub.es.uninstall <- sub.f:
+				break uninstallLoop
+			case <-sub.f.consensusInfo:
+			}
+		}
+
+		// wait for filter to be uninstalled in work loop before returning
+		// this ensures that the manager won't use the event channel which
+		// will probably be closed by the client asap after this method returns.
+		<-sub.Err()
+	})
+}
+
+// subscribe installs the subscription in the event broadcast loop.
+func (es *EventSystem) subscribe(sub *subscription) *Subscription {
+	es.install <- sub
+	<-sub.installed
+	return &Subscription{ID: sub.id, f: sub, es: es}
+}
+
+// SubscribePendingTxs creates a subscription that writes transaction hashes for
+// transactions that enter the transaction pool.
+func (es *EventSystem) SubscribeConsensusInfo(consensusInfo chan *MinimalEpochConsensusInfo, epoch eth2Types.Epoch) *Subscription {
+	sub := &subscription{
+		id:            rpc.NewID(),
+		typ:           MinConsensusInfoSubscription,
+		created:       time.Now(),
+		isNew:         true,
+		epoch:         epoch,
+		consensusInfo: consensusInfo,
+		installed:     make(chan struct{}),
+		err:           make(chan error),
+	}
+	return es.subscribe(sub)
+}
+
+type filterIndex map[Type]map[rpc.ID]*subscription
+
+// handleConsensusInfoEvent
+func (es *EventSystem) handleConsensusInfoEvent(filters filterIndex, ev *MinimalEpochConsensusInfo) {
+	for _, f := range filters[MinConsensusInfoSubscription] {
+		if f.isNew {
+			es.sendConsensusInfo(f, ev)
+			f.isNew = false
+			continue
+		}
+		f.consensusInfo <- ev
+	}
+}
+
+// eventLoop (un)installs filters and processes mux events.
+func (es *EventSystem) eventLoop() {
+	// Ensure all subscriptions get cleaned up
+	defer func() {
+		es.consensusInfoSub.Unsubscribe()
+	}()
+
+	index := make(filterIndex)
+	for i := UnknownSubscription; i < LastIndexSubscription; i++ {
+		index[i] = make(map[rpc.ID]*subscription)
+	}
+
+	for {
+		select {
+		case ev := <-es.consensusInfoCh:
+			es.handleConsensusInfoEvent(index, ev)
+		case f := <-es.install:
+			index[f.typ][f.id] = f
+			close(f.installed)
+		case f := <-es.uninstall:
+			delete(index[f.typ], f.id)
+			close(f.err)
+
+		// System stopped
+		case <-es.consensusInfoSub.Err():
+			return
+		}
+	}
+}
+
+// sendConsensusInfo
+func (es *EventSystem) sendConsensusInfo(f *subscription, ev *MinimalEpochConsensusInfo) {
```

Comment:
This method does not need. When any subscriber is registered, it gets all the previous consensus info from `fromEpoch` to current epoch.

Reply:
Solved. We removed custom event system and manged to use `stateNotifier.StateFeed().Send(&feed.Event)`